### DISCUSSION TITLE
UploadPreview not initially mounted in Popover

### DISCUSSION THREAD

**Discussion by JohannesKaufmann** [rpldy/react-uploady#125] [Q&A] [answered]
I have the Provider and two Components: 
- `Summary` which renders a progress bar for the current batch and a helpful text (e.g. "4 of 10 files uploaded")
- `Queue` which renders the individual files from the batch using `UploadPreview`.

The `Queue` is opened in a popover when the user clicks on the `Summary`.

<img width="381" alt="Screenshot 2021-01-13 at 12 20 24" src="https://user-images.githubusercontent.com/20886667/104446072-35bf8300-559a-11eb-99ed-cc2ba70d38c7.png">

_The Screenshot is taken with the Popover already open when the batch starts. If the popover is initially closed, then `UploadPreview` renders nothing when opened later._

The problem is, that [react-tiny-popover](https://github.com/alexkatz/react-tiny-popover) doesn't [render the content](https://github.com/alexkatz/react-tiny-popover/blob/master/src/Popover.tsx#L222) when the popover is not open.
But the `UploadPreview` needs to listen to all messages.

Is there a way to **replay the messages** or **get the current state** from the context?

---

_Awesome library BTW. I was looking quite a while for something like this (specifically a queue). Finally found the library through the dev.to post 🎉. I searched for "react uploader queue" on google and the "hooked on file uploads with react-uploady" post was the first result._

**Comment by JohannesKaufmann** [comment]
## Summary
```jsx
const getStats = (batch: Batch | null) => {
  if (!batch) return {};
  const { items } = batch;

  const totalFiles = batch.items.length;

  const finishedFiles = items.filter(
    (e) => e.state === FILE_STATES.FINISHED || e.state === FILE_STATES.ERROR
  ).length;

  const isFinished = finishedFiles === totalFiles;

  return {
    isFinished,
    totalFiles,
    finishedFiles,
    progressPercent: batch.completed,
  };
};

const UploaderSummary = () => {
  const [currentBatch, setCurrentBatch] = useState<Batch | null>(null);
  useBatchStartListener((batch) => {
    setCurrentBatch(batch);
  });

  useBatchProgressListener((batch) => {
    setCurrentBatch(batch);
  });

  useBatchFinishListener((batch) => {
    setCurrentBatch(batch);
  });

  const { isFinished, progressPercent, totalFiles, finishedFiles } = getStats(
    currentBatch
  );

  let text;
  if (!currentBatch) {
    text = `no uploads yet`;
  } else if (isFinished) {
    text = `all ${totalFiles} uploaded`;
  } else {
    text = `${finishedFiles} of ${totalFiles} uploaded`;
  }

  const [isOpen, setIsOpen] = useState(true);

  return (
    <Tooltip
      isOpen={isOpen}
      onClickOutside={() => setIsOpen(false)}
      content={
        <div
          style={{
            minWidth: 200,
            minHeight: 400,
          }}
        >
          <UploaderQueue />
        </div>
      }
    >
      <SummaryContainer
        onClick={() => {

          setIsOpen(!isOpen);
        }}
      >
        <SummaryText>{text}</SummaryText>

        <Line
          strokeWidth={1}
          percent={progressPercent}
          trailColor={"lightgray"}
          strokeColor={isFinished ? "#00a626" : "#2db7f5"}
        />
      </SummaryContainer>
    </Tooltip>
  );
};
```

_Tooltip is an abstraction over `Popover` because I initially used another tooltip library_

## Queue


```jsx
const UploaderQueue = () => {
  return (
    <UploadPreview
      rememberPreviousBatches={false}
      PreviewComponent={QueueItem}
      onPreviewsChanged={(previews) => {
        console.log("onPreviewsChanged", previews);
      }}
    />
  );
};
```


## QueueItem
```jsx
  ...

  const [response, setResponse] = useState<any>({});
  const [completed, setCompleted] = useState(0);
  const [state, setState] = useState(FILE_STATES.ADDED);

  const update = useCallback(
    (item: BatchItem) => {
      setCompleted(item.completed);
      setState(item.state);
      setResponse(item.uploadResponse?.data);
    },
    [setCompleted, setState, setResponse]
  );

  useItemStartListener(update, id);
  useItemProgressListener(update, id);
  useItemAbortListener(update, id);
  useItemErrorListener(update, id);
  useItemFinishListener(update, id);

  ...
```

**Comment by yoavniran** [comment, marked as answer]
Most likely UploadPreview isnt suitable for this specific use-case where its not necessarily rendered while upload is taking place.

Its meant to be a simple component for non-sophisticated use-cases (although it can be helpful in [such](https://github.com/rpldy/react-uploady/blob/master/guides/UploadQueue.md) at times).

In this case, youre probably better off storing the uploads status (as you seem to be doing at least partly) in your own state and then passing it on to a Queue component of your own, with a custom (simple) preview component based on the state.

The BatchItem has all of the data you need to render the preview.

Hope this helps.

**Reply by JohannesKaufmann** [reply]
@yoavniran thanks alot! You are totally right, I already had the BatchItems in the State but wasn't passing it along.

Now it works 🎉

For anyone stumbling upon this in the future, here is the code:

```jsx
const UploaderSummary = () => {
  // use `currentBatch` for the Summary and the Queue
  const [currentBatch, setCurrentBatch] = useState<Batch | null>(null);
  useBatchStartListener((batch) => {
    setCurrentBatch(batch);
  });
  useBatchProgressListener(...)
  useBatchFinishListener(...);

   // the currentBatch includes `items` which is all the info we need.
  return (
    <Tooltip
      content={
        <UploaderQueue items={currentBatch?.items} />
      }
    ...
  )
}
```

```jsx
import { BatchItem } from "@rpldy/shared";

interface IProps {
  items?: BatchItem[]
}
const UploaderQueue = ({ items = [] }: IProps) => {
  // here we had `UploadPreview` before
  return (
    <div>
      {items.map((item) => {
        return <QueueItem {...item} />;
      })}
    </div>
  );
}
```

```jsx
import { BatchItem, FILE_STATES } from "@rpldy/shared";

const QueueItem = memo((props: BatchItem) => {
  const { id, completed, state } = props;
  const { name } = props.file;
  const { data } = props.uploadResponse ?? {};

  // -> don't need these anymore
  // const update = useCallback(...)
  // useItemStartListener(update, id);
  // useItemProgressListener(update, id);
  // useItemAbortListener(update, id);
  // useItemErrorListener(update, id);
  // useItemFinishListener(update, id);

  ...
```
